concordium-consensus/tests/scheduler/SchedulerTests/Payday.hs

```diff
@@ -74,6 +107,181 @@ testDoMintingP4 = do
               stoFoundationAccount = foundationAccount
             }
 
+prop_mintAmountsEqNewMint :: MintDistribution 'ChainParametersV1 -> MintRate -> Amount -> Bool
+prop_mintAmountsEqNewMint md mr amt = mintTotal (doCalculatePaydayMintAmounts md mr amt) == mintAmount mr amt
+
+instance Arbitrary (UpdateValue 'ChainParametersV1) where
+  arbitrary = UVMintDistribution <$> arbitrary
+
+prop_mintDistributionMostRecent :: MintDistribution 'ChainParametersV1 -> MintRate -> Slot -> [(Slot, UpdateValue 'ChainParametersV1)] -> Amount -> Bool
+prop_mintDistributionMostRecent md mr ps updates amt =
+  let (UVMintDistribution mostRecentMintDistribution) = snd
+        . maximumBy (\(a, _) (b, _) -> compare a b)
+        . takeWhile ((<= ps) . fst)
+        . filter (\(_, UVMintDistribution _) -> True)
+        $ (0, UVMintDistribution md) : updates
+  in calculatePaydayMintAmounts md mr ps updates amt == doCalculatePaydayMintAmounts mostRecentMintDistribution mr amt
+
+-- reward accounts are only increased by the sum of transaction fees
+prop_mintDistributionImmediate :: forall m. (BlockStateOperations m, TreeStateMonad m, MonadProtocolVersion m)
+    => UpdatableBlockState m
+    -- ^Block state
+    -> BlockPointerType m
+    -- ^Parent block
+    -> Slot
+    -- ^Block slot
+    -> BakerId
+    -- ^Baker ID
+    -> Epoch
+    -- ^Epoch of the new block
+    -> Maybe FinalizerInfo
+    -- ^Info on finalization committee for included record, if any
+    -> SeedState
+    -- ^New seed state
+    -> Amount
+    -- ^Transaction fees
+    -> FreeTransactionCounts
+    -- ^Number of "free" transactions of each type
+    -> [(Slot, UpdateValue (ChainParametersVersionFor (MPV m)))]
+    -- ^Ordered chain updates since the last block
+    -> m Bool
+prop_mintDistributionImmediate bs0 blockParent slotNumber bid newEpoch mfinInfo newSeedState transFees freeCounts updates = do
+  oldChainParameters <- bsoGetChainParameters bs0
+  (mintRewardParams, bs1) <- updateBirkParameters newSeedState bs0 oldChainParameters []
+  rewardsBefore <- (^. bankRewardAccounts) <$> bsoGetBankStatus bs1
+  bs2 <- mintAndReward bs1 blockParent slotNumber bid newEpoch mintRewardParams mfinInfo transFees freeCounts updates
+  rewardsAfter <- (^. bankRewardAccounts) <$> bsoGetBankStatus bs2
+  return $ rewardsBefore == rewardsAfter
+
+prop_transactionFeesDistributionP4 :: forall m. (BlockStateOperations m, MonadProtocolVersion m, ChainParametersVersionFor (MPV m) ~ 'ChainParametersV1, AccountVersionFor (MPV m) ~ 'AccountV1)
```

Comment:
Please describe what this is meant to be testing so that the test is maintainable.